# Time in Music

In [1]:
import time
import sys
import pygame
import os

pygame 2.6.1 (SDL 2.32.56, Python 3.13.9)
Hello from the pygame community. https://www.pygame.org/contribute.html


/opt/anaconda3/envs/workshops/lib/python3.13/site-packages/pygame/pkgdata.py:25: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.
  from pkg_resources import resource_stream, resource_exists


## The `pygame` Module

Python is by *far* not the best programming language for game design, but the `pygame` module still has some useful features for creating user interfaces, including an audio mixer.

In [2]:
## Initialize the mixer:

pygame.mixer.init()

## The Winstons - "Amen, Brother"

In the "Amen" folder, there are a selection of `.wav` files, all individual drum hits from a famous drum solo by Gregory C. Coleman of [The Winstons](https://en.wikipedia.org/wiki/The_Winstons), on a 1969 track called "Amen, Brother".

This mid-song drum solo, or "break beat" (so called because it takes place during a break in other instrumentation) is one of the most-sampled music recordings of all time. Whether or not you're aware of it, you've probably heard it before. It has spawned entire genres of electronic music. 

Neither Gregory Coleman nor any of the other Winstons members ever received royalties for the drum loop's use in *thousands* of published works of music.

* ***If you care to dive down a rabbit hole about music culture, sampling, intellectual property, and the public domain, [Nate Harrison's video essay "Can I Get an Amen?" from 2004](https://www.youtube.com/watch?v=XPoxZW8JzzM) is worth a watch.***

### Drum Samples

In [3]:
kick1 = 'Amen/Amen 1-1 (Kick).wav'
kick2 = 'Amen/Amen 1-2 (Kick).wav'
snare1 = 'Amen/Amen 1-3 (Snare).wav'
hat1 = 'Amen/Amen 1-4a (Hat).wav'
snarelite1 =  'Amen/Amen 1-4b (Snarelite).wav'
hat2 = 'Amen/Amen 1-5a (Hat).wav'
snarelite2 = 'Amen/Amen 1-5b (Snarelite).wav'
kickride = 'Amen/Amen 1-6a (Kickride).wav'
kick3 = 'Amen/Amen 1-6b (Kick).wav'
snare2 = 'Amen/Amen 1-7 (Snare).wav'
hat3 = 'Amen/Amen 1-8a (Hat).wav'
snarelite3 = 'Amen/Amen 1-8b (Snarelite).wav'

In [4]:
pygame.mixer.music.load(snare1)
pygame.mixer.music.play()

### Note Values

If you've never dealt with music theory before, don't worry about this too much, but most music is divided into short segments called "measures" that contain a specified number of beats (regularly-spaced rhythmic pulses). A "4/4" time signature or "Common time" as it is sometimes called represents four beats to a measure, each with a value of a "quarter note". The rhythm is felt in groups of four.

Tempo is a measure of how many beats there are in a given time period. In the original sample from "Amen, Brother", the tempo is a little under 140 beats per minute. We're going to use 140 as a starting point.

In [5]:
# Define beats per minute
bpm = 140

# Define note lengths
half = 30 / bpm
quarter = 60 / bpm
eighth = 60 / bpm / 2
sixteenth = 60 / bpm / 4
thirtysecond = 60 / bpm / 8
sixtyfourth = 60 / bpm / 16

In [6]:
eighth

0.21428571428571427

In [7]:
sixteenth

0.10714285714285714

An eighth note (half of a quarter note, or half of a "beat" in 4/4 time) played at a tempo of 140 beats per minute is ~0.21429 seconds long. A sixteenth note is ~0.10714 seconds. If we know these note values, we can send them to `time.sleep()` to instruct Python to wait after playing a sample for that duration before playing the next sample.

This lets us play the audio samples at a consistent tempo, even accounting for different note values.

In [8]:
def play_sample(sample, duration=eighth):
    pygame.mixer.music.load(sample)
    pygame.mixer.music.play()
    print(sample)
    time.sleep(duration)

In [9]:
def play_measure(samples, durations):
    for sample, duration in zip(samples, durations):
        play_sample(sample, duration)

In [10]:
# 1 measure
# 4 eighth notes and 8 sixteenth notes comprise a full measure... just add up the fractions!
# The snare drums fall on the 2nd and 4th quarter notes of the measure

samples = [
    kick1,
    kick2, 
    snare1, 
    hat1, snarelite1, 
    hat2, snarelite2, 
    kickride, kick3, 
    snare2, 
    hat3, snarelite3
]

durations = [
    eighth, 
    eighth, 
    eighth, 
    sixteenth, sixteenth, 
    sixteenth, sixteenth,
    sixteenth, sixteenth,
    eighth, 
    sixteenth, sixteenth]

In [11]:
for i in range(4):
    play_measure(samples, durations)

Amen/Amen 1-1 (Kick).wav
Amen/Amen 1-2 (Kick).wav
Amen/Amen 1-3 (Snare).wav
Amen/Amen 1-4a (Hat).wav
Amen/Amen 1-4b (Snarelite).wav
Amen/Amen 1-5a (Hat).wav
Amen/Amen 1-5b (Snarelite).wav
Amen/Amen 1-6a (Kickride).wav
Amen/Amen 1-6b (Kick).wav
Amen/Amen 1-7 (Snare).wav
Amen/Amen 1-8a (Hat).wav
Amen/Amen 1-8b (Snarelite).wav
Amen/Amen 1-1 (Kick).wav
Amen/Amen 1-2 (Kick).wav
Amen/Amen 1-3 (Snare).wav
Amen/Amen 1-4a (Hat).wav
Amen/Amen 1-4b (Snarelite).wav
Amen/Amen 1-5a (Hat).wav
Amen/Amen 1-5b (Snarelite).wav
Amen/Amen 1-6a (Kickride).wav
Amen/Amen 1-6b (Kick).wav
Amen/Amen 1-7 (Snare).wav
Amen/Amen 1-8a (Hat).wav
Amen/Amen 1-8b (Snarelite).wav
Amen/Amen 1-1 (Kick).wav
Amen/Amen 1-2 (Kick).wav
Amen/Amen 1-3 (Snare).wav
Amen/Amen 1-4a (Hat).wav
Amen/Amen 1-4b (Snarelite).wav
Amen/Amen 1-5a (Hat).wav
Amen/Amen 1-5b (Snarelite).wav
Amen/Amen 1-6a (Kickride).wav
Amen/Amen 1-6b (Kick).wav
Amen/Amen 1-7 (Snare).wav
Amen/Amen 1-8a (Hat).wav
Amen/Amen 1-8b (Snarelite).wav
Amen/Amen 1-1 (Kick).w

### "Swing Time"

To change the rhythmic feel, or "groove", you can mimic swing time or swung rhythm by changing the times of sixteenth notes depending on where they fall in the measure. A sixteenth note that lines up with eighth-note subdivisions will be slightly longer in duration, but a sixteenth note that falls off of an eighth-note subdivision will have a shorter duration.

In [12]:
swing = 0.15

sixteenth * (1+swing)+ sixteenth * (1-swing)

0.21428571428571425

In [13]:
eighth

0.21428571428571427

In [14]:
# 1 measure

swing = 0.15

samples = [
    kick1,
    kick2, 
    snare1, 
    hat1, snarelite1, 
    hat2, snarelite2, 
    kickride, kick3, 
    snare2, 
    hat3, snarelite3
]

durations = [
    eighth, 
    eighth, 
    eighth, 
    sixteenth * (1+swing), sixteenth * (1-swing), 
    sixteenth * (1+swing), sixteenth * (1-swing),
    sixteenth * (1+swing), sixteenth * (1-swing),
    eighth, 
    sixteenth * (1+swing), sixteenth * (1-swing)]

In [15]:
for i in range(4):
    play_measure(samples, durations)

Amen/Amen 1-1 (Kick).wav
Amen/Amen 1-2 (Kick).wav
Amen/Amen 1-3 (Snare).wav
Amen/Amen 1-4a (Hat).wav
Amen/Amen 1-4b (Snarelite).wav
Amen/Amen 1-5a (Hat).wav
Amen/Amen 1-5b (Snarelite).wav
Amen/Amen 1-6a (Kickride).wav
Amen/Amen 1-6b (Kick).wav
Amen/Amen 1-7 (Snare).wav
Amen/Amen 1-8a (Hat).wav
Amen/Amen 1-8b (Snarelite).wav
Amen/Amen 1-1 (Kick).wav
Amen/Amen 1-2 (Kick).wav
Amen/Amen 1-3 (Snare).wav
Amen/Amen 1-4a (Hat).wav
Amen/Amen 1-4b (Snarelite).wav
Amen/Amen 1-5a (Hat).wav
Amen/Amen 1-5b (Snarelite).wav
Amen/Amen 1-6a (Kickride).wav
Amen/Amen 1-6b (Kick).wav
Amen/Amen 1-7 (Snare).wav
Amen/Amen 1-8a (Hat).wav
Amen/Amen 1-8b (Snarelite).wav
Amen/Amen 1-1 (Kick).wav
Amen/Amen 1-2 (Kick).wav
Amen/Amen 1-3 (Snare).wav
Amen/Amen 1-4a (Hat).wav
Amen/Amen 1-4b (Snarelite).wav
Amen/Amen 1-5a (Hat).wav
Amen/Amen 1-5b (Snarelite).wav
Amen/Amen 1-6a (Kickride).wav
Amen/Amen 1-6b (Kick).wav
Amen/Amen 1-7 (Snare).wav
Amen/Amen 1-8a (Hat).wav
Amen/Amen 1-8b (Snarelite).wav
Amen/Amen 1-1 (Kick).w

# Jungle/Drum'n'Bass

The Winstons' drum sample became the basis for several genres of electronic music by virtue of its rhythmic feel and distinct timbre. In the early 1990s, the sample was used extensivley by music producers in the Carribean and the U.K. in fast-paced dance music that chopped up the beat and played with its syncopation.

In [16]:
# Define beats per minute
bpm = 176

# Define note lengths
half = 30 / bpm
quarter = 60 / bpm
eighth = 60 / bpm / 2
sixteenth = 60 / bpm / 4
thirtysecond = 60 / bpm / 8
sixtyfourth = 60 / bpm / 16

In [17]:
# Measures 1 & 3

samples1 = [
    kick1,
    hat1, hat2,
    snare1, 
    kick1,
    hat1, snarelite1,
    snare1,
    kick1, 
    kick2
]

durations1 = [
    eighth, 
    sixteenth, sixteenth,
    eighth, 
    eighth, 
    sixteenth, sixteenth,
    eighth, 
    eighth, 
    eighth]

In [18]:
# Measure 2

samples2 = [
    kick1,
    hat1, hat2,
    snare1, 
    kick1,
    hat1, snarelite1,
    snare1,
    kick1, snarelite1,
    snare2
]

durations2 = [
    eighth, 
    sixteenth, sixteenth,
    eighth, 
    eighth, 
    sixteenth, sixteenth,
    eighth, 
    sixteenth, sixteenth, 
    eighth]

In [19]:
# Measure 4

samples3 = [
    kick1,
    hat1, hat2,
    snare1, 
    hat1, snarelite1,
    hat2, snarelite2,
    kick1,
    snare1, snare2,
    snare1, snare2
]

durations3 = [
    eighth, 
    sixteenth, sixteenth,
    eighth, 
    sixteenth, sixteenth, 
    sixteenth, sixteenth,
    eighth, 
    sixteenth, sixteenth, 
    sixteenth, sixteenth]

In [20]:
play_measure(samples1, durations1)
play_measure(samples2, durations2)
play_measure(samples1, durations1)
play_measure(samples3, durations3)
play_sample(kick1, eighth)

Amen/Amen 1-1 (Kick).wav
Amen/Amen 1-4a (Hat).wav
Amen/Amen 1-5a (Hat).wav
Amen/Amen 1-3 (Snare).wav
Amen/Amen 1-1 (Kick).wav
Amen/Amen 1-4a (Hat).wav
Amen/Amen 1-4b (Snarelite).wav
Amen/Amen 1-3 (Snare).wav
Amen/Amen 1-1 (Kick).wav
Amen/Amen 1-2 (Kick).wav
Amen/Amen 1-1 (Kick).wav
Amen/Amen 1-4a (Hat).wav
Amen/Amen 1-5a (Hat).wav
Amen/Amen 1-3 (Snare).wav
Amen/Amen 1-1 (Kick).wav
Amen/Amen 1-4a (Hat).wav
Amen/Amen 1-4b (Snarelite).wav
Amen/Amen 1-3 (Snare).wav
Amen/Amen 1-1 (Kick).wav
Amen/Amen 1-4b (Snarelite).wav
Amen/Amen 1-7 (Snare).wav
Amen/Amen 1-1 (Kick).wav
Amen/Amen 1-4a (Hat).wav
Amen/Amen 1-5a (Hat).wav
Amen/Amen 1-3 (Snare).wav
Amen/Amen 1-1 (Kick).wav
Amen/Amen 1-4a (Hat).wav
Amen/Amen 1-4b (Snarelite).wav
Amen/Amen 1-3 (Snare).wav
Amen/Amen 1-1 (Kick).wav
Amen/Amen 1-2 (Kick).wav
Amen/Amen 1-1 (Kick).wav
Amen/Amen 1-4a (Hat).wav
Amen/Amen 1-5a (Hat).wav
Amen/Amen 1-3 (Snare).wav
Amen/Amen 1-4a (Hat).wav
Amen/Amen 1-4b (Snarelite).wav
Amen/Amen 1-5a (Hat).wav
Amen/Amen 1-

Using dedicated Digital Audio Workstation software like GarageBand, FL Studio, Reason, Cubase, Live, Logic, or even ProTools is likely an easier route to producing music, but if you really like building things from scratch, Python can work, too!